src/emulator/apphosting/environments.ts

```diff
@@ -0,0 +1,95 @@
+import { readFileFromDirectory, wrappedSafeLoad } from "../../utils";
+import { pathExists } from "fs-extra";
+import { join } from "path";
+import { logger } from "./utils";
+import { Emulators } from "../types";
+
+type EnvironmentAvailability = "BUILD" | "RUNTIME";
+
+const APPHOSTING_YAML = "apphosting.yaml";
+const APPHOSTING_LOCAL_YAML = "apphosting.local.yaml";
+
+// Schema of apphosting.*.yaml files
+interface AppHostingYaml {
+  env?: {
+    variable: string;
+    secret?: string;
+    value?: string;
+    availability?: EnvironmentAvailability[];
+  }[];
+}
+
+interface AppHostingConfiguration {
+  environmentVariables?: { [key: string]: string };
+  secrets?: { [key: string]: string };
+}
+
+/**
+ * Reads an apphosting.*.yaml file, parses, and returns in an easy to use
+ * format.
+ */
+export async function loadAppHostingYaml(
+  sourceDirectory: string,
+  fileName: string,
+): Promise<AppHostingConfiguration> {
+  const file = await readFileFromDirectory(sourceDirectory, fileName);
+  const apphostingYaml: AppHostingYaml = await wrappedSafeLoad(file.source);
+
+  const environmentVariables: { [key: string]: string } = {};
+  const secrets: { [key: string]: string } = {};
+
+  if (apphostingYaml.env) {
+    apphostingYaml.env.map((env) => {
+      if (env.value) {
+        environmentVariables[env.variable] = env.value;
+      }
+
+      if (env.secret) {
+        secrets[env.variable] = env.secret;
+      }
+    });
+  }
+
+  return { environmentVariables, secrets };
+}
+
+/**
+ * Loads in apphosting.yaml & apphosting.local.yaml, giving
+ * apphosting.local.yaml precedence if present.
+ */
+export async function getLocalAppHostingConfiguration(
```

Comment:
FYI: I wrote a very similar code for loading .env files few years ago.

https://github.com/firebase/firebase-tools/blob/48f08ce0d7a29101c6ef717cf71d8563e157a824/src/functions/env.ts#L366

interesting (for me) to see how we approached the problem slightly differently.